oidc/clientassertion/client_assertion.go

```diff
@@ -0,0 +1,206 @@
+	// for JWT claims
+	clientID string
+	audience []string
+	headers  map[string]string
+
+	// for signer
+	alg jose.SignatureAlgorithm
+	// key may be any key type that jose.SigningKey accepts for its Key
+	key any
+	// secret may be used instead of key
+	secret string
+
+	// these are overwritten for testing
+	genID func() (string, error)
+	now   func() time.Time
+}
+
+// Validate validates the expected fields
+func (c *ClientAssertion) Validate() error {
+	var errs []error
+	if c.genID == nil {
+		errs = append(errs, ErrMissingFuncIDGenerator)
+	}
+	if c.now == nil {
+		errs = append(errs, ErrMissingFuncNow)
+	}
+	// bail early if any internal func errors
+	if len(errs) > 0 {
+		return errors.Join(errs...)
+	}
+
+	if c.clientID == "" {
+		errs = append(errs, ErrMissingClientID)
+	}
+	if len(c.audience) == 0 {
+		errs = append(errs, ErrMissingAudience)
+	}
+	if c.alg == "" {
+		errs = append(errs, ErrMissingAlgorithm)
+	}
+	if c.key == nil && c.secret == "" {
+		errs = append(errs, ErrMissingKeyOrSecret)
+	}
+	if c.key != nil && c.secret != "" {
+		errs = append(errs, ErrBothKeyAndSecret)
+	}
+	return errors.Join(errs...)
+}
+
+// SignedToken returns a signed JWT in the compact serialization format
+func (c *ClientAssertion) SignedToken() (string, error) {
+	if err := c.Validate(); err != nil {
+		return "", err
+	}
+	builder, err := c.builder()
+	if err != nil {
+		return "", err
+	}
+	token, err := builder.CompactSerialize()
+	if err != nil {
+		return "", fmt.Errorf("failed to sign token: %w", err)
+	}
+	return token, nil
+}
+
+func (c *ClientAssertion) builder() (jwt.Builder, error) {
+	signer, err := c.signer()
+	if err != nil {
+		return nil, err
+	}
+	id, err := c.genID()
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate token id: %w", err)
+	}
+	claims := c.claims(id)
+	return jwt.Signed(signer).Claims(claims), nil
+}
+
+func (c *ClientAssertion) signer() (jose.Signer, error) {
+	sKey := jose.SigningKey{
+		Algorithm: c.alg,
+	}
+
+	// Validate() ensures these are mutually exclusive
+	if c.secret != "" {
+		sKey.Key = []byte(c.secret)
+	}
+	if c.key != nil {
+		sKey.Key = c.key
+	}
+
+	sOpts := &jose.SignerOptions{
+		ExtraHeaders: make(map[jose.HeaderKey]interface{}, len(c.headers)),
+	}
+	// note: extra headers can override "kid"
+	for k, v := range c.headers {
+		sOpts.ExtraHeaders[jose.HeaderKey(k)] = v
+	}
+
+	signer, err := jose.NewSigner(sKey, sOpts.WithType("JWT"))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create jwt signer: %w", err)
+	}
+	return signer, nil
+}
+
+func (c *ClientAssertion) claims(id string) *jwt.Claims {
+	now := c.now().UTC()
+	return &jwt.Claims{
+		Issuer:    c.clientID,
+		Subject:   c.clientID,
+		Audience:  c.audience,
+		Expiry:    jwt.NewNumericDate(now.Add(5 * time.Minute)),
+		NotBefore: jwt.NewNumericDate(now.Add(-1 * time.Second)),
+		IssuedAt:  jwt.NewNumericDate(now),
+		ID:        id,
+	}
+}
+
+// Options configure the ClientAssertion
+type Option func(*ClientAssertion)
```

Comment:
In Boundary nowadays we usually prefer the format
```suggestion
type Option func(context.Context, *ClientAssertion) error
```
For option types, so that errors can happen in the option, and context for cancellation/values. The context isn't necessary, but the error is nice, that allows validation to happen in the option, which can then propagate up through the function that consumes the option.

Reply:
I dig this. I was following the `oidc` package convention, but an `error` here could save some logic in `Validate`.